core/src/main/scala/zio/schema/Schema.scala

```diff
@@ -41,6 +41,13 @@ object Schema {
 
   final case class ListSchema[A](codec: Schema[A]) extends Schema[List[A]]
 
+  final case class Case[A, Z](id: String, codec: Schema[A], construct: A => Z, deconstruct: Z => Option[Z])
```

Comment:
I love this, as it should minimize boxing! The type signature is a bit off, however, it should be:

```suggestion
  final case class Case[A, Z](id: String, codec: Schema[A], construct: A => Z, deconstruct: Z => Option[A])
```

(i.e. `Option[A]` in the return value of `deconstruct`).

Second, I wonder if we can do a _bit_ better from a performance standpoint.

What if we require that `A <: Z`? Then, because `A` is already a subtype of `Z`, we get `construct` for free!

That means we don't have to use an identity function or tuple arguments to feed into a case class constructor.

That only leaves the `Option` allocation in `deconstruct`. We could eliminate that, if we wanted, but we would need a low-level, dirty trick that we would have to try to hide:

```scala
final case class Case[A, Z](id: String, codec: Schema[A], construct: A => Z, unsafeDeconstruct: Z => A) {
  def deconstruct(z: Z): Option[A] = try { Some(unsafeDeconstruct(z)) } catch { case _ : IllegalArgumentException => None }
}
```

where we would engineer the low-level `unsafeDeconstruct` to throw an `IllegalArgumentException` if the specified `Z` is in fact not the required subtype `A`.

What do you think?

Reply:
Not sure I fully get the deconstruct change. Is it just so that we can avoid the allocation in the codec decoders? There were can "look up" the case by index or id so we can just call `unsafeDeconstruct` directly. But wouldn't we incur a penalty in the encoders? When we do this:

```scala
    private def enumEncoder[Z](cases: Schema.Case[_, Z]*): JsonEncoder[Z] =
      (value: Z, indent: Option[Int], out: Write) => {
        var matched = false
        cases
          .map(c => c -> c.deconstruct(value))
          .foreach {
            case (c, Some(cast)) =>
              matched = true
              out.write('{')
              val indent_ = bump(indent)
              pad(indent_, out)
              string.unsafeEncode(JsonFieldEncoder.string.unsafeEncodeField(c.id), indent_, out)
              if (indent.isEmpty) out.write(':')
              else out.write(" : ")
              schemaEncoder(c.codec.asInstanceOf[Schema[Any]]).unsafeEncode(cast, indent, out)
              out.write('}')
            case _ => ()
          }
        if (!matched) out.write("{}")
      }
```

for each case that doesn't match we're still allocating the `IllegalArgumentException` and also incurring the exception handling overhead. We could of course use `unsafeDeconstruct` directly but then we are saving the Option allocation on the matching case but also incurring the exception overhead on all the others.

Reply:
Nvm. I see now. We can remove the second allocation by calling unsafeDeconstruct in the Protobuf encoder after we find the case and need to cast before serialization. Added a couple of microbenchmarks and it is ~20-25% higher throughput. 

Made the other changes as well.